### DISCUSSION TITLE
Code improvement wherever possible

### DISCUSSION THREAD

**Discussion by stefanos82** [aristocratos/btop#434] [Development]
I have tried to find real-time contact information such as Discord or IRC channel but could not find anything, therefore I will ask here if it's alright.

I have noticed some coding "peculiarities" as I like to call them that could be replaced with improved code.

Is it OK to fork, make the necessary changes, and then PR or do I need to follow a certain procedure first?

Please advise.

Thank you very much.

**Comment by aristocratos** [comment]
@stefanos82
No procedures needed, just follow the guidelines and formatting specified in [CONTRIBUTING.md](https://github.com/aristocratos/btop/blob/main/CONTRIBUTING.md).

Note that btop++ was the project I used to learn C++, so there's likely to be a lot of inconsistencies in the efficiency of the code (and a lot of opportunities for optimization).

If the coding "peculiarities" you mention is about style and not function you would also need a good argument to why the code should be changed (that's not based on just personal opinion).

**Reply by stefanos82** [reply]
> If the coding "peculiarities" you mention is about style and not function you would also need a good argument to why the code should be changed (that's not based on just personal opinion).

There are parts of the code, such as C-style type casts, that could lead to undefined behavior if at any future time you would want to re-cast your return code to something else via `reinterpret_cast`; so, my suggestion would be to improve it just a bit.

Here's an example: https://github.com/aristocratos/btop/blob/ee0945969d46c17a291142b600173a11068e390d/src/btop_tools.cpp#L254-L264

Since you are using C++20 to compile the code, you could do the following: `return string{str_v};` which is list-initializing your `string` without temporary created since C++17.

Another thing that it's good to know is for the following sample: https://github.com/aristocratos/btop/blob/ee0945969d46c17a291142b600173a11068e390d/src/btop_shared.hpp#L233-L248

Personally I would have taken an advantage of the C++ Core Standards and go with the following syntax:

```c++
//* Container for process information
struct proc_info {
	size_t pid{};           // defaults to 0
	string name{};          // defaults to ""
	string cmd{};           // defaults to ""
	string short_cmd{};     // defaults to ""
	size_t threads{};       // defaults to 0
	int name_offset{};      // defaults to 0
	string user{};          // defaults to ""
	uint64_t mem{};         // defaults to 0
	double cpu_p{};         // defaults to 0.0
	double cpu_c{};         // defaults to 0.0
	char state = '0';
	uint64_t p_nice{};      // defaults to 0
	uint64_t ppid{};        // defaults to 0
	uint64_t cpu_s{};       // defaults to 0
	uint64_t cpu_t{};       // defaults to 0
	string prefix{};        // defaults to ""
	size_t depth{};         // defaults to 0
	size_t tree_index{};    // defaults to 0
	bool collapsed{};       // defaults to false
	bool filtered{};        // defaults to false
};
```

Last by not least, the following code https://github.com/aristocratos/btop/blob/ee0945969d46c17a291142b600173a11068e390d/src/btop_tools.hpp#L82-L95

unnecessarily takes a `const int&` as parameter, but clearly that is not what the actual member method is expecting for input:

![image](https://user-images.githubusercontent.com/1777466/193423657-9ccb48b1-c1d7-47fc-9a66-9c1e0a7fa3f0.png)

 I hope this helped a bit...

Cheers and keep up the good work :+1:

**Reply by aristocratos** [reply]
All really good points!

A lot of the functions in btop_tools was written pretty early on where I just learnt that passing const refs to functions was "faster"... mistakes were made :)

Had no idea that default constructing int, double, bool would set them to 0, 0.0, false. Also looks a lot cleaner.
I should probably take the time to actually read through https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines

I've been planning to go back and refactor, there's a lot of code that I would have written differently today (and some "experiments" with bitmasks, etc. that just make the code hard to read without any actual noticeable performance benefits...).
But I haven't really had time yet.

You are more than welcome to submit PR's for fixing mistakes (or just poorly written code).
Anyone willing to help out is very much welcome since I'm the only one actively working on this project.

**Reply by stefanos82** [reply]
Very nice! :+1: Expect Pull Requests then :smiley:

**Reply by stefanos82** [reply]
Currently working on [my local fork](https://github.com/stefanos82/btop) to finish the whole cleanup and I have noticed something that I wanted to ask your opinion @aristocratos 

Do you think we could add a dependency that will help to clean the code, especially the huge concatenated strings?

I'm talking about using [fmt](https://github.com/fmtlib/fmt).

Since `std::format()` is not fully implemented on all compilers that support C++20, I personally use the aforementioned library to format my strings and helps a lot.

Let me share an example to make a point by sharing my point of view:

```c++
out += Mv::to(y, x+9) + title_left + (not filter_text.empty() ? Fx::b : "") + Theme::c("hi_fg") + 'f'
				+ Theme::c("title") + (not filter_text.empty() ? ' ' + filter_text : "ilter")
				+ (not filtering and not filter_text.empty() ? Theme::c("hi_fg") + " del" : "")
				+ (filtering ? Theme::c("hi_fg") + ' ' + Symbols::enter : "") + Fx::ub + title_right;
```

The concatenated spaghetti above could get cleaned up a lot with the following logic:

```c++
out = fmt::format("My username is {} and I am {} years old!", username, age);
fmt::print("{}\n", out);
```

I used to use IOStreams a lot, but lately I switched to `fmt` and have found my inner peace :laughing: when you witness the line reduction, you will feel the same (I hope!).

**Reply by aristocratos** [reply]
I started writing btop++ just as C++20 was being finalized and was a bit annoyed that `std::format` wasn't included in GCC's C++20 implementation, because that would have made it so much simpler.

But yes, as long the `fmt` library works well (and compiles well) on the supported OS's, I'm all for it.
Would also make some of the text formatting functions in btop_tools.cpp redundant.

It is however a pretty big undertaking, since there's a lot of big strings all over the place with many (sometimes nested) ternary operators sprinkled in, so can be quite a handful to untangle :grimacing:

**Reply by aristocratos** [reply]
I believe there could also be something gained by creating references at the start of the function to the colors that are usually re-used throughout the strings instead  of calling `Theme::c()` all the time.
I'm guessing those function calls are probably optimized away so like no performance benefits, but could make it easier to read perhaps.

**Reply by stefanos82** [reply]
True, that's why it's a long-term task that should be not taken lightly; it will take time and effort but will worth it.

**Reply by stefanos82** [reply]
> I believe there could also be something gained by creating references at the start of the function to the colors that are usually re-used throughout the strings instead of calling `Theme::c()` all the time. I'm guessing those function calls are probably optimized away so like no performance benefits, but could make it easier to read perhaps.

Maybe clearer names for each color, much like those `fmt` is using: https://fmt.dev/latest/api.html#terminal-color-and-text-style 

Yeah, the compiler does a lot of magic behind the scenes via optimization(s), which helps a lot with performance as well even though the compilation times could increase a bit; still though the end result is way better now than what we have had with C++98.

**Reply by aristocratos** [reply]
> Maybe clearer names for each color, much like those fmt is using: https://fmt.dev/latest/api.html#terminal-color-and-text-style

Considering the names used now are the names for the theme defined 24-bit colors. `"main_fg"` being the main foreground text color, so not sure what clarity would be gained by not matching the same names as in the .theme files.

What I meant, was at the start of functions that use a lot of the same theme colors, put something like:
```
const auto& main_fg = Theme::c("main_fg); //Main foreground color
const auto& hi_fg = Theme::c("hi_fg"); //Highlight foreground color
const auto& hi_bg = Theme::c("hi_bg"); //Highlight background color
```
and so on, to get rid of all the function calls in the strings.

**Reply by stefanos82** [reply]
Ah OK, got it.

**Reply by stefanos82** [reply]
> What I meant, was at the start of functions that use a lot of the same theme colors, put something like:
> 
> ```
> const auto& main_fg = Theme::c("main_fg); //Main foreground color
> const auto& hi_fg = Theme::c("hi_fg"); //Highlight foreground color
> const auto& hi_bg = Theme::c("hi_bg"); //Highlight background color
> ```
> 
> and so on, to get rid of all the function calls in the strings.

What I wanted to ask is why did you even created the inline function anyway?

You could achieve the same result with `auto hi_bg = Theme::colors["hi_bg"];` and save an unnecessary function call.

I guess you didn't notice that during the initial time of the original implementation.

**Reply by aristocratos** [reply]
Not sure actually, I believe the thinking at the time might have been to at the start of functions use `using Theme;` and then just call `c("main_fg")` to shorten it, which it turns out I forgot about...

**Reply by stefanos82** [reply]
Should we introduce a nested namespace in `Theme` named `Color` so we can include the color names in there?

Something like `Theme::Color::hi_bg` so we can use namespace alias: `namespace tc = Theme::Color;` thus to cleaner usage, like `tc::hi_bg;`, but of course a better name would make a lot more sense than `tc`.

**Comment by stefanos82** [comment]
By the way, since GCC 12.1 we can use `-fuse-ld=` flag to choose a different linker when we want to.

I personally use [mold](https://github.com/rui314/mold) combined with [ccache](https://ccache.dev/) which improves compilation-times sometimes by orders of magnitude!

Any feedback on that?

**Reply by aristocratos** [reply]
Well, compile times haven't really been any issue in this project, pretty sure most x86 systems from the last decade or so are gonna build btop in under 1 minute.

Anyone wanting to build with non default linker can just set the desired flags when calling `make`, so no issues there really.

**Reply by stefanos82** [reply]
Oh that's the usage of `ADDFLAGS`, very clever! :smiley:

**Comment by stefanos82** [comment]
Currently I'm working on replacing `const bool&` to `bool`.

Reminder: Boolean literals `false` and `true` are `prvalue`s of type `bool`, therefore we trust the compiler to do its magics with optimization(s).

Question: What exactly is this boolean variable doing in `btop_draw.cpp`, named `show_temp`? I don't understand :thinking: , is your intention to use it as `1` in case it's `true`?

```c++
		//* Calculate and draw cpu box outlines
		if (Cpu::shown) {
			using namespace Cpu;
            bool show_temp = (Config::getB("check_temp") and got_sensors);
			width = round((double)Term::width * width_p / 100);
			height = max(8, (int)ceil((double)Term::height * (trim(boxes) == "cpu" ? 100 : height_p) / 100));
			x = 1;
			y = cpu_bottom ? Term::height - height + 1 : 1;

			b_columns = max(1, (int)ceil((double)(Shared::coreCount + 1) / (height - 5)));
			if (b_columns * (21 + 12 * show_temp) < width - (width / 3)) {
				b_column_size = 2;
				b_width = (21 + 12 * show_temp) * b_columns - (b_columns - 1);
			}
			else if (b_columns * (15 + 6 * show_temp) < width - (width / 3)) {
				b_column_size = 1;
				b_width = (15 + 6 * show_temp) * b_columns - (b_columns - 1);
			}
			else if (b_columns * (8 + 6 * show_temp) < width - (width / 3)) {
				b_column_size = 0;
			}
			else {
				b_columns = (width - width / 3) / (8 + 6 * show_temp);
				b_column_size = 0;
			}

			if (b_column_size == 0) b_width = (8 + 6 * show_temp) * b_columns + 1;
			b_height = min(height - 2, (int)ceil((double)Shared::coreCount / b_columns) + 4);
                        ...
                        ...
```

**Reply by aristocratos** [reply]
> Question: What exactly is this boolean variable doing in btop_draw.cpp, named show_temp? I don't understand 🤔 , is your intention to use it as 1 in case it's true?

Haven't seen that before? I guess it's more common in python.
Instead of an ternary statement just use the bool to either negate or keep the values depending on if `true (1)` or `false (0)`.

**Reply by stefanos82** [reply]
I guess 

```c++
bool show_temp = (Config::getB("check_temp") and got_sensors);
```

should be replaced with 

```c++
int show_temp = (Config::getB("check_temp") and got_sensors) ? 1 : 0;
```

to make it more obvious of what it supposed to do.

**Reply by aristocratos** [reply]
Or just a comment to explain the behavior, I thought this usage of bools was common knowledge.

**Reply by stefanos82** [reply]
I can honestly say I have never seen such code in C++, thus my surprise when I saw it.

**Comment by stefanos82** [comment]
I'm currently working on `btop_theme.cpp` and we have the following code:

```c++
namespace {
	//* Convert 24-bit colors to 256 colors
	int truecolor_to_256(const int& r, const int& g, const int& b) {
		//? Use upper 232-255 greyscale values if the downscaled red, green and blue are the same value
		if (const int red = round((double)r / 11); red == round((double)g / 11) and red == round((double)b / 11)) {
			return 232 + red;
		}
		//? Else use 6x6x6 color cube to calculate approximate colors
		else {
			return round((double)r / 51) * 36 + round((double)g / 51) * 6 + round((double)b / 51) + 16;
		}
	}
}
```

I have an idea how to get rid of the C-style type casting and I want your feedback.

Here's how the code will look:

```c++
namespace {
	//* Convert 24-bit colors to 256 colors
	int truecolor_to_256(const int& r, const int& g, const int& b) {
		//? Use upper 232-255 greyscale values if the downscaled red, green and blue are the same value
		if (const int red = round((r+0.0) / 11); red == round((g+0.0) / 11) and red == round((b+0.0) / 11)) {
			return 232 + red;
		}
		//? Else use 6x6x6 color cube to calculate approximate colors
		else {
			return round((r+0.0) / 51) * 36 + round((g+0.0) / 51) * 6 + round((b+0.0) / 51) + 16;
		}
	}
}
```

As you can see, we are taking an advantage of implicit conversion mechanism and since we are using `round()`, we know it expects either of the following types: `float`, `double`, `long double`, or `IntegralType` (I hope I see well here https://en.cppreference.com/w/cpp/numeric/math/round).

So, what do you think?

**Reply by aristocratos** [reply]
Sure, that works. And easily tested to give the same results: https://godbolt.org/z/s3rvxzbKK

But I'm not entirely sure why people are so hung up on C-style casts. Like here, where it's from a `int` to a `double`, I'm not sure I see why it should matter.

**Reply by stefanos82** [reply]
I personally like to use any modern C++ type cast **only** when it's necessary; if the compiler offers me a different way to convert to a type I'm interested in, then I prefer that method.

As I have said, C-style type casts when combined with C++'s `reinterpret_cast` could lead to trouble which is something we want to avoid at all costs.

**Reply by aristocratos** [reply]
> As I have said, C-style type casts when combined with C++'s reinterpret_cast could lead to trouble which is something we want to avoid at all costs.

Sure, but that wouldn't apply here since the cast is to a temporary created inside `round()` so the cast is never captured by a variable that can be `reinterpret_cast`ed.

**Reply by stefanos82** [reply]
Fair enough; I won't apply those changes then.

**Reply by aristocratos** [reply]
Well to me it doesn't really matter as long as it works 😉 
I'm just arguing that it wouldn't give better readability compared to the C-style cast and there is no UB to guard against (that I can see) in this particular case.

**Reply by stefanos82** [reply]
By the way, I have tested 3 separate scenarios of casting a variable and produced the same assembly output:

```c++
int main()
{
    int i = 11;
    // return static_cast<float>(i);
    // return i+0.0f;
    return (float)i; 
}
```
Assembly output:

```asm
main:
        push    rbp
        mov     rbp, rsp
        mov     DWORD PTR [rbp-4], 11
        pxor    xmm0, xmm0
        cvtsi2ss        xmm0, DWORD PTR [rbp-4]
        cvttss2si       eax, xmm0
        pop     rbp
        ret
```

Pay close attention in C++'s comment section of `f` suffix; that little detail produces two different assembly commands when we forget to add it to `0.0` _in case we are dealing with **float values** and not double_:

```c++
int main()
{
    int i = 11;
    return i+0.0;
}
```

Assembly Output:

```asm
main:
        push    rbp
        mov     rbp, rsp
        mov     DWORD PTR [rbp-4], 11
        pxor    xmm0, xmm0
        cvtsi2sd        xmm0, DWORD PTR [rbp-4]
        cvttsd2si       eax, xmm0
        pop     rbp
        ret
```

It converts it to a double-precision floating-point value whereas the previous cases are of single-precision floating-point value; tiny detail yet so important!

I learned something new today :smiley: 

P.S.: Don't get fooled that C-style typecasting and C++ typecasting work the same, just because they produced the same assembly output; prefer C++'s typecasting when needed as it has more compile-time checking mechanisms than C's.
